### PR TITLE
feat: Create building block prompts using inquirer in fe-fpm-writer and create ui package to visualize propts from fe-fpm-writer

### PR DESCRIPTION
**Pull request by 815are**
### Value
#2194

#### @sap-ux/fe-fpm-writer
New API class [PromptsAPI](https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/fe-fpm-writer/src/prompts/api.ts) which supports following public methods:
1. `getPrompts` - returns questions for passed `PromptType`. Additionally and optionally returns:
1.2.  `groups` if prompt type has such;
1.2. `initialAnswers` to support case when there no intention to render/ask question, but end generation expects value/answer. In case of building blocks it is property -> [buildingBlockData.buildingBlockType](https://github.com/SAP/open-ux-tools/blob/main/packages/fe-fpm-writer/src/building-block/types.ts#L20)
2. `getChoices` - method to request dynamic choices when UI asks/request it;
3. `validateAnswers` - method to validate passed answers against validator method in questions;
4. `submitAnswers` - method submits passed answers against mapped generator function;
5. `getCodeSnippets` - method returns potential code snippet(s) for passed answers;

Currently PromptsAPI supports generation for 3 types building blocks - `Chart`, `Table`, `FilterBar`

#### @sap-ux/ui-prompting
New ui package which can be used to render question based on call from `@sap-ux/fe-fpm-writer` `PromptsAPI.getPrompts`. Some points to mention:
1. Main component is [Questions](https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Questions/Questions.tsx)
2. storybook support, which can be launched using `pnpm storybook`
3. Currently 3 type of questions supported -> `list`(Combobox), `checkbox`(Multi select Combobox), `input`(TextField input)

image

Additional small changes:
1. `@sap-ux/ui-components` -> export interface for UISelectableOption(ISelectableOption from fluent-ui) which support generic type for "data" property
2. `@sap-ux/project-access` -> new public/exported method "getCapServiceName" which returns CAP service name by passed datasource uri

#### `examples/prompting-ui`
`examples/prompting-ui` which consumes both -  `fe-fpm-writer` Prompt API and `ui-prompting`:
1. Open in terminal `examples\prompting-ui`;
9. Run `pnpm storybook`;
10. Test `Chart`, `Table`, `FilterBar` building block questions and submit to generate building block;

Additional features of `examples/prompting-ui`:
1. Path to project can be changed in bottom panel under 'Project path':
image

3. Potential generated code snippet and current answers are visible in bottom panel under 'Code preview'
image

#### `examples/fe-fpm-cli`
`examples/fe-fpm-cli` which consumes both -  `fe-fpm-writer` Prompt API using CLI `inquirer.prompt`:
1. Open in terminal `examples\fe-fpm-cli`;
2. Run `pnpm start`;
4. Go through asked questions using terminal;

### CODE REVIEW

**Comment by 815are** [comment]
Points:
- [x] remove prop dependantPromptNames  from -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L9
+ https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/MultiSelect/MultiSelect.tsx#L10
- [ ] https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L58 + https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L34
proposal:
```
export interface ListPromptQuestionCreationProps {
    placeholder?: string;
}

/**
 * Represents a question prompt for list question with dropdown visualization.
 * Combines properties of ListQuestion and BasePromptQuestion.
 */
export interface ListPromptQuestion<T extends Answers = Answers> extends ListQuestion<T>, ListBasePromptQuestion {
    /**
     * Dependant prompt names which should be updated after value change of current prompt.
     */
    dependantPromptNames?: string[];

    /**
     * Properties of input field for creation mode.
     */
    creation?: boolean | ListPromptQuestionCreationProps;
}
```
- [ ] delete? -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L70
- [ ] missing tests `onChangeCreateSelect ` for -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L42-L46
- [ ] remove sort from -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L75
and apply sort in api where it is necessary
- [ ] https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/utilities/hooks.ts#L41 -> resole data: choice type. Andis
- [ ] refactor. onChange https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/MultiSelect/MultiSelect.tsx#L48-L55 Andis
- [ ] refactor onChange -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L87-L93
- [ ] consume annotation api instead of tgz
- [ ] recheck missing translations
- [ ] do not pass absolute path in viewOrFragmentPath choices
- [ ] restructure generic prompts code out of building blocks folder Andis
- [ ] check if we can avoid post processing convertion Andis
- [ ] do not create fe-fpm-cli folder when storybook is launched in examples\prompting-ui


Unit tests:
- [ ] packages/ui-prompting/scripts/prepare-storybook.ts - Andis
- [x] packages/project-access/src/project/cap.ts - Alina
- [ ] packages/fe-fpm-writer/src/building-block/prompts/api.ts validateAnswers  - Alina
- [x] enable skipped tests and fix -> getCAPServiceChoices  -> [join('app/incidents')]: {

**Comment by 815are** [comment]
Points:
- [x] remove prop dependantPromptNames  from -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L9
https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/MultiSelect/MultiSelect.tsx#L10
- [x] https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L58 + https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L34
proposal: - **Alina**
```
export interface ListPromptQuestionCreationProps {
    placeholder?: string;
}

/**
 * Represents a question prompt for list question with dropdown visualization.
 * Combines properties of ListQuestion and BasePromptQuestion.
 */
export interface ListPromptQuestion<T extends Answers = Answers> extends ListQuestion<T>, ListBasePromptQuestion {
    /**
     * Dependant prompt names which should be updated after value change of current prompt.
     */
    dependantPromptNames?: string[];

    /**
     * Properties of input field for creation mode.
     */
    creation?: boolean | ListPromptQuestionCreationProps;
}
```
- [x] delete 'onClick'? -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L70   **Alina**
- [ ] missing tests `onChangeCreateSelect ` for -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L42-L46
- [x] remove sort from -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L75
and apply sort in api where it is necessary - **Alina**
- [ ] https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/utilities/hooks.ts#L41 -> resole data: choice type. **Andis**
- [ ] refactor. onChange https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/MultiSelect/MultiSelect.tsx#L48-L55 **Andis**
- [ ] refactor onChange -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L87-L93 **Alina**
- [x] consume annotation api instead of tgz **Alina**
- [x] recheck missing translations **Alina**
- [x] do not pass absolute path in viewOrFragmentPath choices e1b5844270da0c2b2a14e7a8d880c278b8827e87
- [ ] restructure generic prompts code out of building blocks folder **Andis**
- [ ] check if we can avoid post processing convertion **Andis**
- [ ] do not create fe-fpm-cli folder when storybook is launched in examples\prompting-ui
- [x] init i18n -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/fe-fpm-writer/src/building-block/prompts/questions/chart.ts#L31  **Alina**


Unit tests:
- [x] packages/ui-prompting/scripts/prepare-storybook.ts - **Andis**
- [x] packages/project-access/src/project/cap.ts - **Alina**
- [x] packages/fe-fpm-writer/src/building-block/prompts/api.ts validateAnswers  - **Alina**
- [x] enable skipped tests and fix -> getCAPServiceChoices  -> [join('app/incidents')]: {


Bugs:
- [x] filterBarId resolution issue:
image

**Comment by 815are** [comment]
Points:
- [x] remove prop dependantPromptNames  from -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L9
https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/MultiSelect/MultiSelect.tsx#L10
- [x] https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L58 + https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L34
proposal: - **Alina**
```
export interface ListPromptQuestionCreationProps {
    placeholder?: string;
}

/**
 * Represents a question prompt for list question with dropdown visualization.
 * Combines properties of ListQuestion and BasePromptQuestion.
 */
export interface ListPromptQuestion<T extends Answers = Answers> extends ListQuestion<T>, ListBasePromptQuestion {
    /**
     * Dependant prompt names which should be updated after value change of current prompt.
     */
    dependantPromptNames?: string[];

    /**
     * Properties of input field for creation mode.
     */
    creation?: boolean | ListPromptQuestionCreationProps;
}
```
- [x] delete 'onClick'? -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L70   **Alina**
- [ ] missing tests `onChangeCreateSelect ` for -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L42-L46
- [x] remove sort from -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L75
and apply sort in api where it is necessary - **Alina**
- [ ] https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/utilities/hooks.ts#L41 -> resole data: choice type. **Andis**
- [ ] refactor. onChange https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/MultiSelect/MultiSelect.tsx#L48-L55 **Andis**
- [x] refactor onChange -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L87-L93 **Alina**
- [x] consume annotation api instead of tgz **Alina**
- [x] recheck missing translations **Alina**
- [x] do not pass absolute path in viewOrFragmentPath choices e1b5844270da0c2b2a14e7a8d880c278b8827e87
- [ ] restructure generic prompts code out of building blocks folder **Andis**
- [ ] check if we can avoid post processing convertion **Andis**
- [x] do not create fe-fpm-cli folder when storybook is launched in examples\prompting-ui
- [x] init i18n -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/fe-fpm-writer/src/building-block/prompts/questions/chart.ts#L31  **Alina**
- [x] Make `name` property in question interface required and adjust code where `name ?? ''` is used **Alina**
- [ ] Support `export type PromptListChoices = ReadonlyArray<DistinctChoice>` instead `export type PromptListChoices = ReadonlyArray<ChoiceOptions>`?  **Alina**
- [ ] Refactor handling of default values in UI. Currently handled in `examples/prompting-ui`, but should be handled in `ui-prompting` **Andis**


Unit tests:
- [x] packages/ui-prompting/scripts/prepare-storybook.ts - **Andis**
- [x] packages/project-access/src/project/cap.ts - **Alina**
- [x] packages/fe-fpm-writer/src/building-block/prompts/api.ts validateAnswers  - **Alina**
- [x] enable skipped tests and fix -> getCAPServiceChoices  -> [join('app/incidents')]: {


Bugs:
- [x] filterBarId resolution issue:
image
- [x] Selection/combobox input.
1. select any value;
3. Enter manually random value, which does not exist in dropdown list.
Problem - input reset value to empty

https://github.com/SAP/open-ux-tools/assets/90789422/22eb6f2b-416a-430c-8237-21cfc30cb3c7

**Comment by 815are** [comment]
Points:
- [x] remove prop dependantPromptNames  from -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L9
https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/MultiSelect/MultiSelect.tsx#L10
- [x] https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L58 + https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L34
proposal: - **Alina**
```
export interface ListPromptQuestionCreationProps {
    placeholder?: string;
}

/**
 * Represents a question prompt for list question with dropdown visualization.
 * Combines properties of ListQuestion and BasePromptQuestion.
 */
export interface ListPromptQuestion<T extends Answers = Answers> extends ListQuestion<T>, ListBasePromptQuestion {
    /**
     * Dependant prompt names which should be updated after value change of current prompt.
     */
    dependantPromptNames?: string[];

    /**
     * Properties of input field for creation mode.
     */
    creation?: boolean | ListPromptQuestionCreationProps;
}
```
- [x] delete 'onClick'? -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L70   **Alina**
- [ ] missing tests `onChangeCreateSelect ` for -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L42-L46
- [x] remove sort from -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L75
and apply sort in api where it is necessary - **Alina**
- [ ] https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/utilities/hooks.ts#L41 -> resole data: choice type. **Andis**
- [ ] refactor. onChange https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/MultiSelect/MultiSelect.tsx#L48-L55 **Andis**
- [x] refactor onChange -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/ui-prompting/src/components/Inputs/Select/Select.tsx#L87-L93 **Alina**
- [x] consume annotation api instead of tgz **Alina**
- [x] recheck missing translations **Alina**
- [x] do not pass absolute path in viewOrFragmentPath choices e1b5844270da0c2b2a14e7a8d880c278b8827e87
- [ ] restructure generic prompts code out of building blocks folder **Andis**
- [ ] check if we can avoid post processing convertion **Andis**
- [x] do not create fe-fpm-cli folder when storybook is launched in examples\prompting-ui
- [x] init i18n -> https://github.com/SAP/open-ux-tools/blob/feat/fe-fpm-writer/add-building-block-ui/packages/fe-fpm-writer/src/building-block/prompts/questions/chart.ts#L31  **Alina**
- [x] Make `name` property in question interface required and adjust code where `name ?? ''` is used **Alina**
- [x] Support `export type PromptListChoices = ReadonlyArray<DistinctChoice>` instead `export type PromptListChoices = ReadonlyArray<ChoiceOptions>`?  **Alina**
- [ ] Refactor handling of default values in UI. Currently handled in `examples/prompting-ui`, but should be handled in `ui-prompting` **Andis**


Unit tests:
- [x] packages/ui-prompting/scripts/prepare-storybook.ts - **Andis**
- [x] packages/project-access/src/project/cap.ts - **Alina**
- [x] packages/fe-fpm-writer/src/building-block/prompts/api.ts validateAnswers  - **Alina**
- [x] enable skipped tests and fix -> getCAPServiceChoices  -> [join('app/incidents')]: {


Bugs:
- [x] filterBarId resolution issue:
image
- [x] Selection/combobox input.
1. select any value;
3. Enter manually random value, which does not exist in dropdown list.
Problem - input reset value to empty

https://github.com/SAP/open-ux-tools/assets/90789422/22eb6f2b-416a-430c-8237-21cfc30cb3c7

File: packages/fe-fpm-writer/src/building-block/prompts/types.ts
Comment (tobiasqueck) [inline packages/fe-fpm-writer/src/building-block/prompts/types.ts:1]: Looks like you forgot deleting an empty file

**Review by tobiasqueck — DISMISSED** [review]
* changeset ok
* despite it being a huge PR, it is easy to understand and follow the code and the coding structure
* types are enhancing the inquire-common types (which was important for me) and APIs are similar to the other prompting modules that are used in generators (but enhanced for your more complex use cases) 
* adding the two projects to examples is great, it made it very easy to test locally
* the only thing that I would "complain" is that instead of creating an example CLI, you could have added the new functionality to the `create` module, however, that can happen in a later PR since this one is already quite a beast

File: packages/fe-fpm-writer/src/building-block/prompts/types.ts
Comment (815are) [reply packages/fe-fpm-writer/src/building-block/prompts/types.ts:1]: thanks, remove with commit - https://github.com/SAP/open-ux-tools/pull/1564/commits/2731381e78d6c160e29d871355be3675c9cf3266

**Review by tobiasqueck — APPROVED** [review]

**Review by Klaus-Keller — DISMISSED** [review]
Thanks @815are, great work!
- code to add prompting UI that can be used by several consumers looks good
- changesets exist
- test coverage is good
- did a local test with storybook

Approved from my side

**Review by Klaus-Keller — APPROVED** [review]
reapproved after lockfile update

**Pull request merged**